k2/csrc/rnnt_decode.cu

```diff
@@ -0,0 +1,732 @@
+  NVTX_RANGE(K2_FUNC);
+  // return directlly if already detached or no frames decoded.
+  if (!attached_ || prev_frames_.empty()) return;
+  std::vector<Ragged<int64_t>> states;
+  std::vector<Ragged<double>> scores;
+  Unstack(states_, 0, &states);
+  Unstack(scores_, 0, &scores);
+  K2_CHECK_EQ(static_cast<int32_t>(states.size()), num_streams_);
+  K2_CHECK_EQ(static_cast<int32_t>(scores.size()), num_streams_);
+
+  // detatch prev_frames_
+  std::vector<Ragged<ArcInfo> *> frames_ptr;
+  for (size_t i = 0; i < prev_frames_.size(); ++i) {
+    frames_ptr.emplace_back(prev_frames_[i].get());
+  }
+  // statck_frames has a shape of [t][stream][state][arc]
+  auto stack_frames = Stack(0, prev_frames_.size(), frames_ptr.data());
+  // stack_frames now has a shape of [strams][state][arc]
+  // its Dim0(0) equals to `num_streams_ * prev_frames_.size()`
+  stack_frames = stack_frames.RemoveAxis(0);
+
+  std::vector<Ragged<ArcInfo>> frames;
+  Unstack(stack_frames, 0, &frames);
+
+  K2_CHECK_EQ(num_streams_ * prev_frames_.size(), frames.size());
+
+  for (int32_t i = 0; i < num_streams_; ++i) {
+    for (size_t j = 0; j < prev_frames_.size(); ++j) {
+      srcs_[i]->prev_frames.emplace_back(
+          std::make_shared<Ragged<ArcInfo>>(frames[j * num_streams_ + i]));
+    }
+    srcs_[i]->states = states[i];
+    srcs_[i]->scores = scores[i];
+  }
+
+  attached_ = false;
+  prev_frames_.clear();
+}
+
+void RnntDecodingStreams::GetContexts(RaggedShape *shape,
+                                      Array2<int32_t> *contexts) {
+  NVTX_RANGE(K2_FUNC);
+  K2_CHECK(shape);
+  K2_CHECK(contexts);
+  K2_CHECK_EQ(states_.NumAxes(), 3);
+
+  // shape has a shape of [stream][context]
+  *shape = RemoveAxis(states_.shape, 2);
+  int32_t num_contexts = shape->TotSize(1),
+          decoder_history_len = config_.decoder_history_len,
+          vocab_size = config_.vocab_size;
+
+  // contexts has a shape of [num_contexts][decoder_history_len]
+  *contexts = Array2<int32_t>(c_, num_contexts, decoder_history_len);
+  const int32_t *shape_row_ids1_data = shape->RowIds(1).Data(),
+                *states_row_splits2_data = states_.RowSplits(2).Data();
+  auto contexts_acc = contexts->Accessor();
+
+  const int32_t *num_graph_states_data = num_graph_states_.Data();
+  const int64_t *states_values_data = states_.values.Data();
+
+  K2_EVAL2(
+      c_, num_contexts, decoder_history_len, lambda_set_contexts,
+      (int32_t row, int32_t col) {
+        int32_t idx0 = shape_row_ids1_data[row],
+                num_graph_states = num_graph_states_data[idx0],
+                state_idx01x = states_row_splits2_data[row];
+        int64_t state_value = states_values_data[state_idx01x];
+        int32_t exp_v = decoder_history_len - col;
+        int64_t context_state = state_value / num_graph_states,
+                state = context_state % (int64_t)pow(vocab_size, exp_v);
+        state = state / pow(vocab_size, exp_v - 1);
+        contexts_acc(row, col) = state;
```

Comment:
It a bit difficult to follow without any documentation.

Reply:
Fixed.